client/libp2p/client/message_sender.go

```diff
@@ -0,0 +1,229 @@
+type peerMessageSender struct {
+	s  network.Stream
+	r  msgio.ReadCloser
+	lk ctxMutex
+	p  peer.ID
+	m  *messageSenderImpl
+
+	invalid bool
+}
+
+// invalidate is called before this peerMessageSender is removed from the strmap.
+// It prevents the peerMessageSender from being reused/reinitialized and then
+// forgotten (leaving the stream open).
+func (ms *peerMessageSender) invalidate() {
+	ms.invalid = true
+	if ms.s != nil {
+		_ = ms.s.Reset()
+		ms.s = nil
+	}
+}
+
+func (ms *peerMessageSender) prepOrInvalidate(ctx context.Context) error {
+	if err := ms.lk.Lock(ctx); err != nil {
+		return err
+	}
+	defer ms.lk.Unlock()
+
+	if err := ms.prep(ctx); err != nil {
+		ms.invalidate()
+		return err
+	}
+	return nil
+}
+
+func (ms *peerMessageSender) prep(ctx context.Context) error {
+	if ms.invalid {
+		return fmt.Errorf("message sender has been invalidated")
+	}
+	if ms.s != nil {
+		return nil
+	}
+
+	// We only want to speak to peers using our primary protocols. We do not want to query any peer that only speaks
+	// one of the secondary "server" protocols that we happen to support (e.g. older nodes that we can respond to for
+	// backwards compatibility reasons).
+	nstr, err := ms.m.host.NewStream(ctx, ms.p, ms.m.protocols...)
+	if err != nil {
+		return err
+	}
+
+	ms.r = msgio.NewVarintReaderSize(nstr, network.MessageSizeMax)
+	ms.s = nstr
+
+	return nil
+}
+
+func (ms *peerMessageSender) sendMessage(ctx context.Context, pmes *pb.Message) error {
+	if err := ms.lk.Lock(ctx); err != nil {
+		return err
+	}
+	defer ms.lk.Unlock()
+
+	if err := ms.prep(ctx); err != nil {
+		return err
+	}
+
+	if err := ms.writeMsg(pmes); err != nil {
+		_ = ms.s.Reset()
+		ms.s = nil
+
+		log.Debugw("error writing message", "error", err, "retrying", true)
+		return err
+	}
+
+	return nil
+}
+
+// SendMessage to a peer without waiting for any response.
+func (ms *peerMessageSender) SendMessage(ctx context.Context, pmes *pb.Message) error {
+	return ms.sendMessage(ctx, pmes)
+}
+
+// SendRequest sends a message and waits for a resonse to be sent back.
+func (ms *peerMessageSender) SendRequest(ctx context.Context, pmes *pb.Message) (*pb.Message, error) {
+	err := ms.sendMessage(ctx, pmes)
+	if err != nil {
+		log.Debugw("Error sending request", "erorr", err)
+		return nil, err
+	}
+
+	mes := new(pb.Message)
+	if err := ms.ctxReadMsg(ctx, mes); err != nil {
+		_ = ms.s.Reset()
+		ms.s = nil
+
+		log.Debugw("error reading message", "error", err)
+		return nil, err
+	}
+
+	return mes, nil
+}
+
+func (ms *peerMessageSender) writeMsg(pmes *pb.Message) error {
+	return net.WriteMsg(ms.s, pmes)
+}
+
+func (ms *peerMessageSender) ctxReadMsg(ctx context.Context, mes *pb.Message) error {
+	errc := make(chan error, 1)
+	go func(r msgio.ReadCloser) {
+		defer close(errc)
+		bytes, err := r.ReadMsg()
+		defer r.ReleaseMsg(bytes)
+		if err != nil {
+			errc <- err
+			return
+		}
+		errc <- mes.Unmarshal(bytes)
+	}(ms.r)
```

Comment:
Instead of starting a new goroutine for every read, consider having a read goroutine that is started when the message sender is created.  The read goroutine would wait on a channel that takes a request object.  The request object contains a `*pb.Message` and an error chan to respond on.